### PR TITLE
`Iris`: Move rate limit into admin settings

### PR DESCRIPTION
**Pull request by Hialus**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The rate-limit introduced in #7246 should be moved out of the application.yml and into the IrisSettings.

### Description

This PR does just that. Most of the changes were previously reviewed in #7246, until I removed them to put it into the application.yml

### Steps for Testing

Prerequisites:
- 1 Admin

1. Log in to Artemis
2. Navigate to Global Iris Settings in the Admin area
3. Set the rateLimit and timeframe to something like 5 and 24
4. Send messages until you have reached the limit
5. Change the limit in the settings
6. It should adjust accordingly in the chat

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [iris-sub-settings.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5594/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/iris/settings/iris-sub-settings.model.ts.html) | 100% | ✅ |
| [exercise-chat-widget.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5594/latest/artifact/shared/Coverage-Report-Client-Tests/app/iris/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.ts.html) | 66.66% | ✅ |
| [heartbeat.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5594/latest/artifact/shared/Coverage-Report-Client-Tests/app/iris/heartbeat.service.ts.html) | 79.31% | ✅ |
| [http-session.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5594/latest/artifact/shared/Coverage-Report-Client-Tests/app/iris/http-session.service.ts.html) | 100% | ✅ |
| [iris-sub-settings-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5594/latest/artifact/shared/Coverage-Report-Client-Tests/app/iris/settings/iris-settings-update/iris-sub-settings-update/iris-sub-settings-update.component.ts.html) | 94.11% | ✅ |
| [state-store.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5594/latest/artifact/shared/Coverage-Report-Client-Tests/app/iris/state-store.model.ts.html) | 86.36% | ✅ |
| [state-store.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5594/latest/artifact/shared/Coverage-Report-Client-Tests/app/iris/state-store.service.ts.html) | 92.4% | ✅ |
| [websocket.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5594/latest/artifact/shared/Coverage-Report-Client-Tests/app/iris/websocket.service.ts.html) | 95.65% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [IrisSubSettings.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5594/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.iris.settings/IrisSubSettings.html) | 100% | ✅ |
| [IrisRateLimitService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5594/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.iris/IrisRateLimitService.html) | 100% | ✅ |
| [IrisSettingsService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5594/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.iris/IrisSettingsService.html) | 88% | ✅ |
| [IrisSessionResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5594/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.iris/IrisSessionResource.html) | 82% | ✅ |

### Screenshots

image
image

### CODE REVIEW

File: src/main/webapp/app/iris/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.ts [hunk 1]
```diff
@@ -633,7 +635,9 @@ export class ExerciseChatWidgetComponent implements OnInit, OnDestroy, AfterView
         if (error.status === 403) {
             this.stateStore.dispatch(new ConversationErrorOccurredAction(IrisErrorMessageKey.IRIS_DISABLED));
         } else if (error.status === 429) {
-            this.stateStore.dispatch(new ConversationErrorOccurredAction(IrisErrorMessageKey.RATE_LIMIT_EXCEEDED));
+            const map = new Map<string, any>();
```
Comment (krusche) [inline src/main/webapp/app/iris/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.ts:638]: avoid the use of `any`

File: src/main/webapp/app/iris/state-store.service.ts [hunk 2]
```diff
@@ -227,11 +229,19 @@ export class IrisStateStore implements OnDestroy {
         }
         if (isRateLimitUpdatedAction(action)) {
             const castedAction = action as RateLimitUpdatedAction;
+            const rateLimitInfo = castedAction.rateLimitInfo;
+            let errorMessage: IrisErrorType | null = null;
+            if (rateLimitInfo.rateLimit >= 0 && rateLimitInfo.currentMessageCount >= rateLimitInfo.rateLimit) {
+                errorMessage = errorMessages[IrisErrorMessageKey.RATE_LIMIT_EXCEEDED];
+                errorMessage.paramsMap = new Map<string, any>();
```
Comment (krusche) [inline src/main/webapp/app/iris/state-store.service.ts:236]: avoid any

File: src/main/webapp/app/iris/state-store.service.ts [hunk 3]
```diff
@@ -227,11 +229,19 @@ export class IrisStateStore implements OnDestroy {
         }
         if (isRateLimitUpdatedAction(action)) {
             const castedAction = action as RateLimitUpdatedAction;
+            const rateLimitInfo = castedAction.rateLimitInfo;
+            let errorMessage: IrisErrorType | null = null;
```
Comment (krusche) [inline src/main/webapp/app/iris/state-store.service.ts:233]: we prefer `undefined` over `null`

File: src/main/webapp/app/iris/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.ts (hunk 1 above)
Comment (Hialus) [reply src/main/webapp/app/iris/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.ts:638]: Will be addressed in a separate PR, as this requires a bigger refactoring of the Iris client.

File: src/main/webapp/app/iris/state-store.service.ts (hunk 2 above)
Comment (Hialus) [reply src/main/webapp/app/iris/state-store.service.ts:236]: Will be addressed in a separate PR, as this requires a bigger refactoring of the Iris client.

File: src/main/webapp/app/iris/state-store.service.ts (hunk 3 above)
Comment (Hialus) [reply src/main/webapp/app/iris/state-store.service.ts:233]: Will be addressed in a separate PR, as this requires a bigger refactoring of the Iris client.

**Review by MichaelOwenDyer — APPROVED** [review]
excellent work 💯

**Review by RY997 — CHANGES_REQUESTED** [review]
Tested on TS9 and found one small problem:
To reproduce: After reaching the limit for the first time, change the value of hours and open the chat window again and the number will increase (and exceed)
image

**Review by RY997 — APPROVED** [review]
LGTM and tested on TS9 again, it's the expected behavior.
Good job!

**Pull request merged**